### PR TITLE
Important PennyLane tests are converted to double as benchmarks

### PR DESCRIPTION
**Pull request by AmintorDusko**
**Context:**

The PennyLane test suite currently hits almost all areas of the code base (~100% coverage). Since all areas of the codebase can be tested, these tests are in a good place to collect benchmarking data of the respective tested functions.

Usage:

For running only benchmarks, in the test file test_compare_default_qubit.py, one can call:
```
python -m pytest pennylane/devices/tests/test_compare_default_qubit.py --device=lightning.qubit --benchmark-enable --benchmark-only
```

**Description of the Change:**

This PR will implement the necessary changes to have tests doubling as benchmarks.
The PennyLane test suite will be set by default to disable benchmarks, and benchmarks will be easily enabled with ``--benchmark-enable``.

**Benefits:**

This initial push will enable some initial tracking of possible performance regressions.

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

**Comment by albi3ro** [comment]
How do we get/ inspect the benchmarks if they are run in the CI?

**Comment by AmintorDusko** [comment]
For anyone who might want to inspect the benchmark artifacts:

Click in benchmarks Details:
image

Go for Summary
image

Now go all the way down to explore the Benchmarks' artifacts:
image

**Comment by AmintorDusko** [comment]
> How do we get/ inspect the benchmarks if they are run in the CI?

For those passing by, CIs were moved to another PR.

**Review by albi3ro — APPROVED** [review]
🎉  :)

File: pennylane/devices/tests/test_compare_default_qubit.py [hunk 1]
```diff
@@ -66,8 +66,18 @@ def circuit(theta, phi):
...
         grad = qml.grad(qnode, argnum=[0, 1])
 
-        assert np.allclose(qnode(theta, phi), qnode_def(theta, phi), atol=tol(dev.shots))
-        assert np.allclose(grad(theta, phi), grad_def(theta, phi), atol=tol(dev.shots))
+        def workload():
+            return (
+                qnode(theta, phi),
+                qnode_def(theta, phi),
+                grad(theta, phi),
+                grad_def(theta, phi),
```
Comment (timmysilv) [inline pennylane/devices/tests/test_compare_default_qubit.py:74]: do we want to benchmark the tested device as well as the reference device?

**Review by timmysilv — APPROVED** [review]
just one comment, not a blocker though. looks great!

File: pennylane/devices/tests/test_compare_default_qubit.py (hunk 1 above)
Comment (AmintorDusko) [reply pennylane/devices/tests/test_compare_default_qubit.py:74]: Good question.
Yes. At least for now, this is the plan.
The selection of benchmarks, and possible the approach to do that, will be refined in the next iteration.

**Comment by AmintorDusko** [comment]
Thank you, @albi3ro and @timmysilv!

**Comment by AmintorDusko** [comment]
I will update the changelog and enable auto-merge asap.

**Pull request merged**